pjrt-plugin/iree/integrations/pjrt/common/debugging.cc

```diff
@@ -0,0 +1,166 @@
+
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+namespace iree::pjrt {
+
+//===----------------------------------------------------------------------===//
+// Logger
+//===----------------------------------------------------------------------===//
+
+void Logger::debug(std::string_view message) {
+  std::cerr << "[IREE-PJRT] DEBUG: " << message << std::endl;
+}
+
+void Logger::error(std::string_view message) {
+  std::cerr << "[IREE-PJRT] ERROR: " << message << std::endl;
+}
+
+//===----------------------------------------------------------------------===//
+// ArtifactDumper
+//===----------------------------------------------------------------------===//
+
+ArtifactDumper::~ArtifactDumper() = default;
+
+std::unique_ptr<ArtifactDumper::Transaction>
+ArtifactDumper::CreateTransaction() {
+  return nullptr;
+}
+
+std::string ArtifactDumper::DebugString() { return std::string("disabled"); }
+
+//===----------------------------------------------------------------------===//
+// FilesArtifactDumper
+//===----------------------------------------------------------------------===//
+
+class FilesArtifactDumper::FilesTransaction final
+    : public ArtifactDumper::Transaction {
+ public:
+  FilesTransaction(Logger& logger, std::filesystem::path base_path,
+                   int64_t transaction_id, bool retain_all)
+      : logger_(logger),
+        base_path_(std::move(base_path)),
+        transaction_id_(transaction_id),
+        retain_all_(retain_all) {}
+  ~FilesTransaction() { Retain(); }
+
+  std::optional<std::string> AllocateArtifactPath(std::string_view label,
+                                                  std::string_view extension,
+                                                  int index) override {
+    std::string basename = std::to_string(transaction_id_);
+    basename.append("-");
+    basename.append(label);
+    if (index >= 0) {
+      basename.append(std::to_string(index));
+    }
+    basename.append(".");
+    basename.append(extension);
+
+    auto file_path = base_path_ / basename;
+    written_paths_.push_back(file_path);
+
+    return file_path;
+  }
+
+  void WriteArtifact(std::string_view label, std::string_view extension,
+                     int index, std::string_view contents) override {
+    // Note that the API declares this as optional, but our implementation
+    // always returns a path.
+    auto file_path = AllocateArtifactPath(label, extension, index);
+
+    std::ofstream fout;
+    fout.open(*file_path, std::ofstream::out | std::ofstream::trunc |
+                              std::ofstream::binary);
+    fout.write(contents.data(), contents.size());
+    fout.close();
+
+    if (!fout.good()) {
+      std::string message("I/O error dumping artifact: ");
+      message.append(*file_path);
+      logger_.error(message);
+    }
+  }
+
+  void Retain() override {
+    if (written_paths_.empty()) return;
+
+    std::string message;
```

Comment:
Should this use string stream?

Reply:
I'm going to fix the logger so it can do formatted messages, but just hadn't taken the time to remember how to do that. When I do, will rip through and fix all of these.